File: trie-serialization/trie-serialization.py
"""
    Definition of TrieNode:
    class TrieNode:
    def __init__(self):
    # <key, value>: <Character, TrieNode>
    self.children = collections.OrderedDict()
    """
class Solution:
    
    '''
        @param root: An object of TrieNode, denote the root of the trie.
        This method will be invoked first, you should design your own algorithm
        to serialize a trie which denote by a root node to a string which
        can be easily deserialized by your own "deserialize" method later.
        '''
    def serialize(self, root):
        # Write your code here
        def dfs(parent):
            ans = "<"
            for key in parent.children:
                value = parent.children[key]
                ans += key
                ans += dfs(value)
            ans += ">"
            return ans
        
        return dfs(root)
    
    
    '''
        @param data: A string serialized by your serialize method.
        This method will be invoked second, the argument data is what exactly
        you serialized at method "serialize", that means the data is not given by
        system, it's given by your own serialize method. So the format of data is
        designed by yourself, and deserialize it here as you serialize it in
        "serialize" method.
        '''
    def deserialize(self, data):
        # Write your code here
        def dfs(parent, str):
            le = len(str)
            i = 1
            while i < le:
                if str[i] != '<' and str[i] != '>':
                    child = TrieNode()
                    parent.children[str[i]] = child
                    j = i + 1
                    cnt = 0
                    while j < le:
                        if str[j] == '<':
                            cnt += 1
                        elif str[j] == '>':
                            cnt -= 1
                        if cnt == 0:
                            break
                        j += 1
                    dfs(child, str[i + 1: j + 1])
                    i = j
                i += 1
        
        root = TrieNode()
        dfs(root, data)
    return root

File: trie-serialization/trie-serialization.cc
/**
 * Definition of TrieNode:
 * class TrieNode {
 * public:
 *     TrieNode() {}
 *     map<char, TrieNode*> children;
 * };
 */

class Solution {
public:
    string serialize_dfs(TrieNode* root) {
        string ans = "<";
        for (map<char, TrieNode*>::iterator it = root->children.begin();
             it != root->children.end(); ++it) {
            string tmp = " ";
            tmp[0] = it->first;
            tmp += serialize_dfs(it->second);
            ans += tmp;
        }
        return ans + ">";
    }
    
    /**
     * This method will be invoked first, you should design your own algorithm
     * to serialize a trie which denote by a root node to a string which
     * can be easily deserialized by your own "deserialize" method later.
     */
    string serialize(TrieNode* root) {
        // Write your code here
        return serialize_dfs(root);
    }
    
    void deserialize_dfs(TrieNode* root, string data, char fa = ' ') {
        int len = data.length();
        for (int i = 1; i < len; ++i) {
            if (data[i] != '<' && data[i] != '>') {
                TrieNode *child = new TrieNode();
                root->children[data[i]] = child;
                int j, cnt = 0;
                for (j = i + 1; j < len; ++j) {
                    if (data[j] == '<') ++cnt;
                    else if (data[j] == '>') --cnt;
                    if (cnt == 0) {
                        break;
                    }
                }
                deserialize_dfs(child, data.substr(i + 1, j - i), data[i]);
                i = j;
            }
        }
    }
    
    /**
     * This method will be invoked second, the argument data is what exactly
     * you serialized at method "serialize", that means the data is not given by
     * system, it's given by your own serialize method. So the format of data is
     * designed by yourself, and deserialize it here as you serialize it in
     * "serialize" method.
     */
    TrieNode* deserialize(string data) {
        // Write your code here
        TrieNode *root = new TrieNode();
        deserialize_dfs(root, data);
        return root;
    }
};
